manifests/broker/service.pp

```diff
@@ -23,31 +23,57 @@
   $service_name = 'kafka'
 
   if $service_install {
-    if $::service_provider == 'systemd' {
-      include ::systemd
-
-      file { "${service_name}.service":
-        ensure  => file,
-        path    => "/etc/systemd/system/${service_name}.service",
-        mode    => '0644',
-        content => template('kafka/unit.erb'),
+    case $::service_provider {
+      'systemd': {
+        include ::systemd
+
+        file { "${service_name}.service":
+          ensure  => file,
+          path    => "/etc/systemd/system/${service_name}.service",
+          mode    => '0644',
+          content => template('kafka/unit.erb'),
+        }
+
+        file { "/etc/init.d/${service_name}":
+          ensure => absent,
+        }
+
+        File["${service_name}.service"] ~>
+        Exec['systemctl-daemon-reload'] ->
+        Service[$service_name]
       }
 
-      file { "/etc/init.d/${service_name}":
-        ensure => absent,
+      'upstart': {
+        file { "/etc/init.d/${service_name}":
+          ensure => absent,
+        } ~>
+
+        exec { 'purge_rcd':
+          command     => 'update-rc.d kafka remove',
+          refreshonly => true
+        } ->
+
+        file { "${service_name}.service":
+          ensure  => file,
+          path    => "/etc/init/${service_name}",
```

Comment:
The upstart config needs a `.conf` suffix or else upstart doesn't recognize it as a service config